XTypes type discovery protocol

dennis-adlink opened the pull request:
This PR adds the implementation of the first stage of the XTypes type discovery protocol in Cyclone. This first implementation is not compliant with the DDS XTypes specification in the sense that it currently uses a different `type_information` format (md5 hash of the serialized sertype), and the exchanged types are not a `TypeObject`, but a serialized `ddsi_sertype` object. At a later stage, when the XTypes type system is implemented, we can plug that into this implementation and get a more spec compliant implementation of the type discovery.

This PR includes a partial implementation of the type consistency enforcement QoS policy and a (limited) assignability check for `sertype_default` that is used when matching readers and writers.

A build option `ENABLE_TYPE_DISCOVERY` is added so that type discovery support can be disabled when building Cyclone. By default type discovery is disabled.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+  struct entidx_enum_participant est;
+  struct participant *pp = NULL;
+  bool pp_found = false;
+  entidx_enum_participant_init (&est, gv->entity_index);
```
There is no need to scan the participants:
* if `guid_prefix_empty(guid_prefix)`, one just picks the first (same code, minus the loop).
* otherwise, one can use the fact that all participants have entity id 0x1c1 (there is a name for it, I can't be bothered to look it up right now), and so one can just do a GUID lookup of (effectively) `{guid_prefix,0x1c1}`.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+
+  struct participant *pp = get_typelookup_pp (gv, &tlm->dst_prefix);
+  struct writer *wr = get_builtin_writer (pp, NN_ENTITYID_TL_SVC_BUILTIN_REQUEST_WRITER);
+  if (wr == NULL)
```
So if an arbitrary participant was picked by `get_typelookup_pp` (the first one in the enumeration is a pretty arbitrary one, given how random the ids are) and it doesn't have a writer for requests, does that guarantee no other participant has a writer?

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+
+  type_lookup_request_t request;
+  memcpy (&request.writer_guid, &wr->e.guid, sizeof (wr->e.guid));
+  request.sequence_number = wr->seq;
```
`wr->seq` is an unlocked access. Given that all requests are written by this function and that it always holds `tl_admin_lock` across this access and sending the sample, it obviously is safe as written. But if there's no need to use the writer's sequence number, I'd rather have a request sequence number in the tl admin.

dennis-adlink replied:
Added a `request_seqno` field in `struct tl_meta`

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+
+  GVTRACE (" wr-tl-req "PGUIDFMT" typeid "PTYPEIDFMT"\n", PGUID (wr->e.guid), PTYPEID(*type_id));
+  struct ddsi_tkmap_instance *tk = ddsi_tkmap_lookup_instance_ref (gv->m_tkmap, serdata);
+  write_sample_gc (lookup_thread_state (), NULL, wr, serdata, tk);
```
I would prefer not holding any locks while writing this, or to use a guaranteed non-blocking write operation. That all depends on QoS and stuff like that — currently a `KEEP_ALL` writer isn't guaranteed non-blocking. But we could make it such for this case.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+  }
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+
+  if (types.n > 0)
```
I'd think a reply should be written also if there is no substance in the response: that way the requestor doesn't have to timeout before it can continue.

dennis-adlink replied:
I agree, removed the count > 0 condition here, so it'll always reply to an incoming type lookup request (unless there is no typelookup reply writer, which should of course not happen where the is a request reader).

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+
+void ddsi_tl_handle_request (struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix, struct ddsi_serdata *sample_common)
+{
+  assert (!(sample_common->statusinfo & (NN_STATUSINFO_DISPOSE | NN_STATUSINFO_UNREGISTER)));
```
I haven't checked yet, but in case I forget: if this is an assert, the caller should ensure that this precondition is met before calling this function.

dennis-adlink replied:
I moved this to `q_ddsi_discovery` as a condition for calling `ddsi_tl_handle_request` and `ddsi_tl_handle_request`

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+{
+  struct generic_proxy_endpoint **gpe_match_upd = NULL;
+  uint32_t n_match_upd = 0;
+  assert (!(sample_common->statusinfo & (NN_STATUSINFO_DISPOSE | NN_STATUSINFO_UNREGISTER)));
```
I haven't checked yet, but in case I forget: if this is an assert, the caller should ensure that this precondition is met before calling this function.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+      tlm->state = TL_META_RESOLVED;
+      tlm->sertype = &st->c; // refcounted by register
+
+      gpe_match_upd = ddsrt_malloc (tlm->n_endpoints * sizeof (*gpe_match_upd));
```
This, inside the outer loop, is not going to work quite so well. And I'm tempted to suggest adding some generic list handling (in `ddsrt`) to make this sort of thing a bit less verbose. Something that constructs a linked list of (short) arrays of entries or so. It could be templatised with macros, too 🙂

dennis-adlink replied:
I've created an abstraction (the `tlm_endpoints_*` functions) for the endpoint list in `struct tl_meta`, which is currently implemented as a simple array, but can easily be replaced by another list implementation if required.

eboasson commented on src/core/ddsi/src/q_ddsi_discovery.c:
```diff
@@ -1545,7 +1565,8 @@ int builtins_dqueue_handler (const struct nn_rsample_info *sampleinfo, const str
              sampleinfo->rst->vendor.id[0], sampleinfo->rst->vendor.id[1],
-             PGUID (guid), sampleinfo->seq, statusinfo, d->topic->name, d->topic->type_name,
+             PGUID (guid), sampleinfo->seq, statusinfo,
+             pwr && (pwr->c.xqos->present & QP_TOPIC_NAME) ? pwr->c.xqos->topic_name : "", d->type->type_name,
```
It'd be better to guarantee a topic name is present in all endpoints (I actually think it is already, or else it could easily be done because it were only ever the built-in ones that did not have a name).

eboasson commented on src/core/ddsi/src/q_qosmatch.c:
```diff
@@ -63,73 +66,128 @@ int partitions_match_p (const dds_qos_t *a, const dds_qos_t *b)
 {
+  assert (rd_tlm->sertype != NULL);
+  assert (wr_tlm->sertype != NULL);
+  return rd_tlm->sertype->ops->assignable_from (rd_tlm->sertype, wr_tlm->sertype);
```
`ddsi_sertype_assignable_from`

eboasson commented on src/core/ddsi/src/q_qosmatch.c:
```diff
@@ -63,73 +66,128 @@ int partitions_match_p (const dds_qos_t *a, const dds_qos_t *b)
-  if ((mask & QP_TYPE_NAME) && strcmp (rd->type_name, wr->type_name) != 0)
+  }
+
+  struct tl_meta *rd_tlm = NULL, *wr_tlm = NULL;
```
I'd make a function out of the type checking.

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -2753,12 +2803,20 @@ static bool topickind_qos_match_p_lock (struct entity_common *rd, const dds_qos_
+
+  if (!ret && (rd_type_lookup || wr_type_lookup))
+  {
+    (void) ddsi_tl_request_type (gv, rd_type_lookup ? rd_typeid : wr_typeid);
```
I get why you can do `rd_type_lookup ? rd_typeid : wr_typeid`, but it would be worth a comment.

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -4583,12 +4681,24 @@ static void create_proxy_builtin_endpoints(
       assert (is_builtin_entityid (guid1.entityid, proxypp->vendor));
       if (is_writer_entityid (guid1.entityid))
       {
+        // Replace topic name for each pwr
```
It seems to me that it would make more sense to refactor it a bit and pass topic & type name in as arguments.

dennis-adlink commented on src/core/ddsi/src/ddsi_sertype_default.c:
```diff
@@ -0,0 +1,213 @@
+  return true;
+}
+
+static uint32_t sertype_default_hash (const struct ddsi_sertype *tpcmn)
```
As far as I have found, this is only used as the hash function for `gv->sertypes`. So lookups will use it and I don't think that causes any performance issues, as the resolved `sertype` is stored in the endpoint entities. I'll update this function to use the first 4 bytes of the type id hash.

dennis-adlink commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -0,0 +1,356 @@
+uint32_t ddsi_tl_meta_hash (const struct tl_meta *tl_meta)
+{
+  uint32_t h = 0;
+  h = ddsrt_mh3 (&tl_meta->type_id->hash, TYPEID_HASH_LENGTH, h);
```
In the current implementation the type identifier only consists of a hash of the type descriptor, so you're right that we can use this in the hash table for the type lookup administration. In the xtypes spec the type identifier is a union type and only optionally includes a hash, so at some point we'll need to build a hash value in this function. For now I'll use the type id hash.

dennis-adlink commented:
@eboasson Thanks for your extensive review! In the additional commits on this PR I've addressed most of the issues (I'll give a reply to your comments later). In one of the Travis builds on Windows for my own fork there was a refcounting issue, I'll also fix that.

dennis-adlink commented on src/core/ddsc/CMakeLists.txt:
```diff
@@ -37,8 +37,8 @@ PREPEND(srcs_ddsc "${CMAKE_CURRENT_LIST_DIR}/src"
     dds_whc_builtintopic.c
-    dds_serdata_builtintopic.c
-    dds_sertopic_builtintopic.c
+    dds_serdata_builtintype.c
```
I've reverted this name change in commit ebf6d2f

dennis-adlink commented on src/core/ddsc/include/dds/dds.h:
```diff
@@ -206,6 +206,8 @@ typedef struct dds_builtintopic_endpoint
   char *topic_name;
   char *type_name;
+  unsigned char *type_identifier;
+  uint32_t type_identifier_sz;
```
You're right, I changed it to size_t

dennis-adlink commented on src/core/ddsc/include/dds/dds.h:
```diff
@@ -1071,21 +1074,23 @@ dds_create_topic(
 dds_create_topic_generic (
   dds_entity_t participant,
-  struct ddsi_sertopic **sertopic,
+  const char *name,
```
Added `DDS_HAS_DDSI_SERTYPE` macro

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/ddsi_serdata.h:
```diff
@@ -230,21 +230,21 @@ DDS_EXPORT inline bool ddsi_serdata_to_sample (const struct ddsi_serdata *d, voi
 
-DDS_EXPORT inline bool ddsi_serdata_topicless_to_sample (const struct ddsi_sertopic *topic, const struct ddsi_serdata *d, void *sample, void **bufptr, void *buflim) {
-  return d->ops->topicless_to_sample (topic, d, sample, bufptr, buflim);
+DDS_EXPORT inline bool ddsi_serdata_topicless_to_sample (const struct ddsi_sertype *type, const struct ddsi_serdata *d, void *sample, void **bufptr, void *buflim) {
```
Good point, I've changed all 'topicless' function names to 'untyped'

dennis-adlink commented on src/core/ddsi/src/ddsi_plist.c:
```diff
@@ -1567,6 +1567,7 @@ static const struct piddesc piddesc_omg[] = {
   QP  (PARTITION,                           partition, XQ, XS, XSTOP),
   QP  (TIME_BASED_FILTER,                   time_based_filter, XD),
   QP  (TRANSPORT_PRIORITY,                  transport_priority, Xi),
+  QP  (TYPE_CONSISTENCY_ENFORCEMENT,        type_consistency, XE1, Xb, Xb, Xb, Xb, Xb),
```
Added Xbx3, Xbx4 and Xbx5

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/ddsi_typeid.h:
```diff
@@ -0,0 +1,68 @@
+
+typedef union {
+  unsigned char c[TYPEID_HASH_LENGTH];
+  uint32_t u[4];
```
Yes, the type identifier is in the current implementation a sequence of `uint8_t`, the `uint32_t` member was added just for convenient access for logging the value, but I rewrote that with some macros

dennis-adlink commented on src/core/ddsc/src/dds_domain.c:
```diff
@@ -365,3 +365,47 @@ void dds_write_set_batch (bool enable)
+  struct tl_meta *tlm = ddsi_tl_meta_lookup (&e->m_domain->gv, &type_id);
+  if (tlm == NULL)
+  {
+    rc = DDS_RETCODE_NOT_FOUND;
```
I didn't realize that `NOT_FOUND` is not defined in the spec. I've changed it so that `dds_domain_resolve_type` returns `PRECONDITION_NOT_MET` in case the type identifier is not known in the local type lookup administration or no type request writer could be found, and `NO_DATA` in case resolving the type failed within the specified duration.

dennis-adlink commented on src/core/ddsc/src/dds_domain.c:
```diff
@@ -365,3 +365,47 @@ void dds_write_set_batch (bool enable)
+  {
+    // FIXME: resolve the type information
+    //ddsi_tl_request_type (&e->m_domain->gv, &type_id);
+    rc = DDS_RETCODE_NOT_FOUND;
```
Yes, it definitely needs a timeout parameter, it wasn't there because I didn't finish the implementation yet. For the return code: I think both `NO_DATA` and `TIMEOUT` would fit here, but I agree with you that `TIMEOUT` is the best option because in most cases one expects the resolving to succeed and return a type.

dennis-adlink commented on src/core/ddsc/src/dds_serdata_builtintype.c:
```diff
@@ -42,127 +42,127 @@ static uint32_t hash_guid (const ddsi_guid_t *g)
-  ddsi_xqos_copy(&d->xqos, rd->xqos);
-  set_topic_type_from_sertopic(d, rd->topic);
+  ddsi_xqos_copy (&d->xqos, rd->xqos);
+  set_topic_name_type (d, rd->xqos->topic_name, rd->type->type_name);
```
Makes sense, I changed this so that `ent->xqos` always has the topic name set.

dennis-adlink commented on src/core/ddsc/src/dds_topic.c:
```diff
@@ -152,7 +152,11 @@ static dds_return_t dds_topic_delete (dds_entity *e)
   assert (dds_entity_kind (e->m_parent) == DDS_KIND_PARTICIPANT);
   dds_participant * const pp = (dds_participant *) e->m_parent;
-  ddsi_sertopic_unref (tp->m_stopic);
+  type_identifier_t *type_id = ddsi_typeid_from_sertype (tp->m_stype);
```
Yes, type identifiers can have variable size in the Xtypes spec, e.g. when using array or map collection types it will contain one or possibly two nested type identifiers. To be prepared for that, I've used dynamic allocation in the current implementation. The type hash field of the type identifier has a fixed length, currently 16 bytes and in the Xtypes implementation it will be 14 bytes.

dennis-adlink commented on src/core/ddsc/src/dds_topic.c:
```diff
@@ -152,7 +152,11 @@ static dds_return_t dds_topic_delete (dds_entity *e)
   dds_participant * const pp = (dds_participant *) e->m_parent;
-  ddsi_sertopic_unref (tp->m_stopic);
+  type_identifier_t *type_id = ddsi_typeid_from_sertype (tp->m_stype);
+  ddsi_tl_meta_unref (&pp->m_entity.m_domain->gv, type_id, NULL);
```
Good point, I've changed this so that ref and unref have the same interface.

dennis-adlink commented on src/core/ddsi/src/ddsi_sertype.c:
```diff
@@ -0,0 +1,183 @@
+  *buf = ddsrt_malloc (*sz);
+  *(uint32_t *)(*buf + i) = ddsrt_toBE4u (tnsz);
+  i += sizeof (tnsz);
+  memcpy (*buf + i, tp->type_name, tnsz);
```
Updated this in f29fa4a to use the `pserop` serializer, for both `ddsi_sertype` and `ddsi_sertype_default` (as you mentioned below).

dennis-adlink commented:
@eboasson the refcounting issue I mentioned in my previous comment is fixed in 9767e54, as well as a bug in locking for the condition variable used by `dds_domain_resolve_type`. The builds on Travis are running stable now for my fork. I'll rebase the PR branch after you've re-reviewed the updates. 
Two things I'll add in short term: the comment on the `dds_domain_resolve_type` functions needs to mention the refcounting for sertype that needs be handled by the application, and I'll add preprocessor conditionals for the type lookup functionality, so that it can be excluded from the build.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -38,56 +38,117 @@ size_t typelookup_service_request_nops = sizeof (typelookup_service_request_ops)
-  ddsrt_free (tlm);
+  if (proxy_ep_guid != NULL)
+  {
+    if (!tlm_endpoint_exists (tlm, proxy_ep_guid))
```
This data structure doesn't scale. As long as systems are small, it is not an issue, and it doesn't make sense to delay merging the PR (especially considering that it defaults to being disabled) because of it, but I do think it needs to be noted.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -38,56 +38,117 @@ size_t typelookup_service_request_nops = sizeof (typelookup_service_request_ops)
-  ddsrt_free (tlm->type_id);
-  ddsrt_free (tlm->endpoints);
-  ddsrt_free (tlm);
+  if (proxy_ep_guid != NULL)
```
I don't like this test here. It does make sense to represent references from topic entities by using a null pointer in the GUID in add/remove, but I don't think that should be forwarded all the way to `endpoint_add` and `endpoint_remove`.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -138,124 +198,140 @@ void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_i
-      GVTRACE (" add ep "PGUIDFMT, PGUID (*src_ep_guid));
-    }
-  }
+  tlm_endpoint_add (gv, tlm, proxy_ep_guid);
```
So this call should be conditional on `proxy_ep_guid` being non-null. A few doc comments for the interface exported by this module would be a good idea, too, by the way.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -138,124 +198,140 @@ void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_i
+  GVTRACE (" state %d refc %u", tlm->state, tlm->refc);
   ddsrt_mutex_unlock (&gv->tl_admin_lock);
+
+  if (resolved)
```
Why can't this simply be all within `tl_admin_lock`?

dennis-adlink replied:
Moved the line to within the lock

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -138,124 +198,140 @@ void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_i
+  struct participant *pp;
+  struct writer *wr = NULL;
+  thread_state_awake (lookup_thread_state (), gv);
+  if (guid_prefix_empty (guid_prefix))
```
See the comment about `dst_prefix` where the containing struct type is defined. I'm not so sure it's worth having.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -280,7 +356,7 @@ void ddsi_tl_handle_request (struct ddsi_domaingv *gv, const ddsi_guid_prefix_t
       types.types = ddsrt_realloc (types.types, types.n * sizeof (*types.types));
-      memcpy (&types.types[types.n - 1].type_identifier, tlm->type_id, sizeof (types.types[types.n - 1].type_identifier));
+      types.types[types.n - 1].type_identifier = tlm->type_id;
       tlm->sertype->ops->serialize (tlm->sertype, &sz, &types.types[types.n - 1].type_object.value);
```
Error handling? I mean, it is defined as returning success/failure, and one does need to at least allow for types that do not support it.

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -289,68 +365,118 @@ void ddsi_tl_handle_request (struct ddsi_domaingv *gv, const ddsi_guid_prefix_t
     {
+      bool sertype_new = false;
       GVTRACE (" type "PTYPEIDFMT, PTYPEID (r.type_identifier));
       st = ddsrt_malloc (sizeof (*st));
```
For the time being I am fine with assuming `ddsi_sertype_default` here

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -289,68 +365,118 @@ void ddsi_tl_handle_request (struct ddsi_domaingv *gv, const ddsi_guid_prefix_t
+  ddsrt_free (types.types);
+}
+
+static void ddsi_tl_meta_proxy_endpoint_ref_locked (struct ddsi_domaingv *gv, struct tl_meta *tlm)
```
The name is a bit confusing because the `ref` and `unref` suffix are pretty consistently used for incrementing and decrementing reference counts, and this is not doing that at all.

dennis-adlink replied:
I agree, I've changed it to `ddsi_tl_meta_register_with_proxy_endpoints`, which better describes what the function does and is less confusing.

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -2793,7 +2811,18 @@ static void reader_qos_mismatch (struct reader * rd, dds_qos_policy_id_t reason)
 }
 
-static bool topickind_qos_match_p_lock (struct ddsi_domaingv *gv, struct entity_common *rd, const dds_qos_t *rdqos, const type_identifier_t *rd_typeid, struct entity_common *wr, const dds_qos_t *wrqos, const type_identifier_t *wr_typeid, dds_qos_policy_id_t *reason)
+static bool topickind_qos_match_p_lock (
```
I have been staring at this (as well as `qos_match_p`) and I don't quite like the set of parameters, but I'm not quite sure much can be done to improve the situation: passing a struct in would require the struct to be initialized, &c. So I'm just mentioning it in case you have a good suggestion 🙂.

dennis-adlink replied:
I didn't really find a better solution for this, maybe the planned refactoring of fields in (proxy) entities will help in reducing the number of parameters for these function calls.

eboasson commented on src/core/ddsi/include/dds/ddsi/ddsi_typelookup.h:
```diff
@@ -76,17 +77,17 @@ extern size_t typelookup_service_reply_nops;
 int ddsi_tl_meta_equal (const struct tl_meta *a, const struct tl_meta *b);
 uint32_t ddsi_tl_meta_hash (const struct tl_meta *tl_meta);
-void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const struct ddsi_sertype *type, const ddsi_guid_t *src_ep_guid, const ddsi_guid_prefix_t *dst);
+void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const struct ddsi_sertype *type, const ddsi_guid_t *proxy_ep_guid, const ddsi_guid_prefix_t *dst);
```
I think it would be advisable to have different versions of `meta_ref` and `meta_unref` so that there are fewer null pointers being passed in. For `ref` it really splits into local entities vs proxies, and for `unref` based on type_id or type.

I am fine with these being wrappers that forward it to the current implementation, it's just that I think it'd be clearer at the call sites.

dennis-adlink replied:
For both `ref` and `unref` I've split up the function into a `proxy` and `local` variant; I think it's more clear to use proxy/local this for unref too, instead of having a `type_id` and `type` variant.

dennis-adlink commented on src/core/ddsc/tests/typelookup.c:
```diff
@@ -0,0 +1,341 @@
+  return result;
+}
+
+static void endpoint_info_free (endpoint_info_t *ep_info)
```
I've used a custom data structure to store only the required data, so I think this free function can stay here.

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/q_entity.h:
```diff
@@ -419,8 +421,10 @@ struct proxy_endpoint_common
   type_identifier_t type_id; /* type identifier for for type used by this proxy endpoint */
-  const struct ddsi_sertype * type; /* type */
+#endif
+  const struct ddsi_sertype * type; /* sertype for data this endpoint reads/writes */
```
Yes you're right, I've changed this

dennis-adlink reviewed (commented):
Thanks for reviewing again @eboasson, in commits ac26478 and b53823e I've addressed the issues you mentioned.

dennis-adlink commented on src/core/ddsc/src/dds_domain.c:
```diff
@@ -378,28 +385,50 @@ dds_return_t dds_domain_resolve_type (dds_entity_t entity, unsigned char *type_i
-    rc = DDS_RETCODE_NOT_FOUND;
-    goto failed;
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    ddsrt_mutex_lock (&gv->tl_resolved_lock);
```
I've refactored this so that it's now only using the `gv->tl_admin_lock` mutex. After requesting the type and re-acquiring the lock, there is a test to check if the type is already resolved, as it may have been resolved after the lock was released in `ddsi_tl_request_type` (which is required because when writing the sample we shouldn't hold the lock). In that case there is no need to wait for `gv->tl_resolved_cond` to be triggered. 

In `ddsi_tl_request_type` there is a test `tlm->state == TL_META_NEW` to avoid duplicate requests for a type, so I think that makes it the 'lookup-and-request-if-unknown' function you mean?

dennis-adlink commented on src/core/ddsi/src/ddsi_sertype.c:
```diff
@@ -118,40 +118,38 @@ struct sertype_ser
   struct sertype_ser d;
   size_t srcoff = 0;
-  if (plist_deser_generic_srcoff (&d, src_data, src_sz, &srcoff, false, sertype_ser_ops) < 0)
+  if (plist_deser_generic_srcoff (&d, serdata, sz, &srcoff, DDSRT_ENDIAN != DDSRT_LITTLE_ENDIAN, sertype_ser_ops) < 0)
```
The last parameter of `plist_ser_generic_embeddable` is a bool that indicates big-endian encoding, not bswap y/n, so that's set to `false` because the xtypes spec (7.3.4.5) requires LE encoding for TypeObject serialization. The deser functions have a `bool bswap` parameter, which make it somewhat confusing. So maybe we could change the ser functions to use `bool bswap` as well?

eboasson replied:
> because the xtypes spec (7.3.4.5) requires LE encoding for TypeObject serialization

Dealing with OMG specs is like dealing with Windows: cruel and unusual punishment — but at least the organisation has a good acronym ... So the security spec mandates big-endian encoding (unnecessarily) while the XTypes spec mandates little-endian encoding (equally unnecessarily) even though the original DDSI specification leaves the decision to the sender except in the one case where a standard endianness is required (the perfidious keyhashes (\*)) and mandates that the receiver can handle either. (I admit to not having read all the fine print in the XTypes spec ...)

The last parameter of `plist_ser_generic_embeddable` is a bool that controls native vs big-endian encoding and was added to accomodate the Security spec's foibles, but setting it to `false` doesn't allow for the XTypes spec's insanity because on a big-endian machine that'll give you big-endian. So it needs to be complexified further and changed into a tri-state.

(\*) The keyhash is perfidious because for keys of which the serialized length may be larger than 16 bytes, it uses the MD5 hash, and therefore creating a topic with an octet sequence as key and writing two instances with keys that happen to produce an MD5 hash collision has "interesting" consequences for implementations that rely on the keyhash. And yet the DDS Security spec mandates the presence of the keyhash for those cases where the readers can't decode the data but need to keep instances separate, effectively forbidding applications from choosing the key values freely, but rather requiring them to avoid MD5 collisions. Not much good can be said about that.

eboasson replied:
P.S.: of course a boolean that says whether or not to byteswap while serialising suffices to do any of the modes we need. But that pushes the tests like `DDSRT_ENDIAN != DDSRT_LITTLE_ENDIAN`) to the caller. I think the interface would be easier to use if you could specify the desired result as native/little/big endian and then let the serializer function figure out whether or not it needs to do byteswapping.

dennis-adlink commented:
@eboasson thanks for reviewing again! I've covered most of the remarks from your review in commit 8defdec (see also the commit message), except for the byteswapping, which is not wrong I think, but somewhat confusing because of the parameters of the ser/deser functions.

dennis-adlink commented:
Thanks @eboasson! I agree with your suggestion to replace the assert on type info length with a check, that's fixed in c95520e. The refactoring of the endianness parameter in serialization is in c399890, and finally I fixed some null function pointers that should be 0 in commit 6be9031.

eboasson reviewed (approved):
@dennis-adlink this PR has spent an embarrassingly long time with me and I am grateful that you have been so patient. Today, I am happy to finally declare it suitable for merging, yay!

Still, the actual merging will have to wait for a little longer: because it breaks source compatibility (which is fine, it is just an evolving interface) some updates need to go into other repositories first. In particular:
* `rmw_cyclonedds_cpp`, for which I have got the changes done and have locally checked that it works with "old" and "new" sources, and also that an old `rmw_cyclonedds_cpp` is binary compatible with a "new" `cyclonedds`. The PR for that should show up today or tomorrow.
* `cyclonedds-cxx`, which needs a similar adjustment to what I did for `rmw_cyclonedds_cpp`. Perhaps @reicheratwork would be willing to have a look?
* the Rust binding on https://github.com/atolab (I suspect it uses a pinned version of `cyclonedds`, it which case that can wait).
